mlem/contrib/kubernetes/base.py

```diff
@@ -0,0 +1,255 @@
+    """docker registry"""
+    daemon: Optional[DockerDaemon] = DockerDaemon(host="")
+    """docker daemon"""
+    kube_config_file_path: Optional[str] = None
+    """path for kube config file of the cluster"""
+
+    def load_kube_config(self):
+        config.load_kube_config(
+            config_file=self.kube_config_file_path
+            or os.getenv("KUBECONFIG", default="~/.kube/config")
+        )
+
+    def find_index(self, nodes_list, node_name):
+        for i, each_node in enumerate(nodes_list):
+            if each_node.metadata.name == node_name:
+                return i
+        return -1
+
+    def _get_client(self, state: K8sDeploymentState) -> Client:
+        host, port = None, None
+        self.load_kube_config()
+        service = client.CoreV1Api().list_namespaced_service(self.namespace)
+        try:
+            if self.service_type == "NodePort":
+                port = service.items[0].spec.ports[0].node_port
+                node_name = (
+                    client.CoreV1Api()
+                    .list_namespaced_pod(self.namespace)
+                    .items[0]
+                    .spec.node_name
+                )
+                node_list = client.CoreV1Api().list_node().items
+                node_index = self.find_index(node_list, node_name)
+                address_dict = node_list[node_index].status.addresses
+                for each_address in address_dict:
+                    if each_address.type == "ExternalDNS":
+                        host = each_address.address
+            elif self.service_type == "LoadBalancer":
+                port = service.items[0].spec.ports[0].port
+                ingress = service.items[0].status.load_balancer.ingress[0]
+                host = ingress.hostname or ingress.ip
+            else:
+                raise ValueError(
+                    f"service_type supplied is {self.service_type}, valid values are [ClusterIP, NodePort, LoadBalancer] only"
+                )
+        except (IndexError, ValueError, TypeError) as e:
+            print(
+                "Couldn't determine host and port from the service deployed:",
+                e,
+            )
+        if host is not None and port is not None:
+            return HTTPClient(host=host, port=port)
+        raise TypeError(
+            f"host and port determined are not valid, received host as {host} and port as {port}"
+        )
+
+
+class K8sEnv(MlemEnv[K8sDeployment]):
+    """MlemEnv implementation for Kubernetes Environments"""
+
+    type: ClassVar = "kubernetes"
+
+    deploy_type: ClassVar = K8sDeployment
+    """type of deployment being used for the Kubernetes environment"""
+    registry: Optional[DockerRegistry] = None
+    """docker registry"""
+
+    def get_registry(self, meta: K8sDeployment):
+        if self.registry is None and meta.registry is None:
+            echo(EMOJI_FAIL + "Error: No registry set to be used by Docker")
+            raise ValueError(
```

Comment:
should be mlem error. and you wont need an echo before this

Reply:
done